### ISSUE TITLE
Linux tests hanging

### ISSUE THREAD

**Issue by alfredclwong** [microsoft/OpticSim.jl#62]
Our Linux test runs have been hanging indefinitely since the introduction of the new emitter code. The first test run that revealed this behaviour can be seen [here](https://github.com/microsoft/OpticSim.jl/runs/2180489935). This is the tail end of the output:

```
Test Summary: | Pass  Total
Comparison    |  121    121
[ Info: Makie/AbstractPlotting is caching fonts, this may take a while. Needed only on first run!
Error: The operation was canceled.
```

It seems to be stuck on the Visualization tests, so it's probably a Makie issue. @friggog @BrianGun, do you think it could be something to do with the precompilation / display stack issue that's dealt with by the `__init__() = AbstractPlotting.__init__()` line? @galran did you notice any issues locally when you ran the tests? They're running fine on my machine.

**To Reproduce**
1. Checkout the new emitter code in a branch
2. Trigger a CI.yml run with Ubuntu (any supported Julia version)

**Expected behavior**
After a 6 hour timeout (or cancellation), we see the above log.

**Testing server:**
 - OS: Ubuntu 18.04
 - Julia version: 1.5, 1.6, 1.7
 - OpticSim.jl version: latest (0.3.1 + new emitter code)

**Comment by galran** [comment]
I didn't notice any problems on windows. Didn't try it on a headless Linux machine.

ran

From: Alfred Wong ***@***.***>
Sent: Friday, March 26, 2021 12:33 PM
To: microsoft/OpticSim.jl ***@***.***>
Cc: Ran Gal ***@***.***>; Mention ***@***.***>
Subject: [microsoft/OpticSim.jl] Linux tests hanging (#62)


Our Linux test runs have been hanging indefinitely since the introduction of the new emitter code. The first test run that revealed this behaviour can be seen here<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FOpticSim.jl%2Fruns%2F2203734986%3Fcheck_suite_focus%3Dtrue&data=04%7C01%7Crgal%40microsoft.com%7C5c4bbaa0161f40857f5c08d8f08df17f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637523839716297817%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=uy%2Bilj8n%2BH7omAZQILZsJtOQiTvt43GUi7NF4%2FPXAlY%3D&reserved=0>. This is the tail end of the output:

Test Summary: | Pass  Total

Comparison    |  121    121

[ Info: Makie/AbstractPlotting is caching fonts, this may take a while. Needed only on first run!

Error: The operation was canceled.

It seems to be stuck on the Visualization tests, so it's probably a Makie issue. @friggog<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ffriggog&data=04%7C01%7Crgal%40microsoft.com%7C5c4bbaa0161f40857f5c08d8f08df17f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637523839716307822%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Li0CvxUBS%2Bs%2FN5veC3%2B0aJ37to0X%2BNes6BPQDmcmU%2FQ%3D&reserved=0> @BrianGun<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FBrianGun&data=04%7C01%7Crgal%40microsoft.com%7C5c4bbaa0161f40857f5c08d8f08df17f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637523839716307822%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=wj8DdvDP2r05h5MK%2F%2ForQhf9GrQqwmKFwX7Cy3A78MI%3D&reserved=0>, do you think it could be something to do with the precompilation / display stack issue that's dealt with by the __init__() = AbstractPlotting.__init__() line? @galran<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgalran&data=04%7C01%7Crgal%40microsoft.com%7C5c4bbaa0161f40857f5c08d8f08df17f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637523839716317813%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=OtshdBgHNLpKJQQB8IY2dRnSpMWgKxs3EaPrpImwZu4%3D&reserved=0> did you notice any issues locally when you ran the tests? They're running fine on my machine.

To Reproduce
Steps to reproduce the behavior:

  1.  Checkout the new emitter code in a branch
  2.  Trigger a CI.yml run with Ubuntu (any supported Julia version)

Expected behavior
After a 6 hour timeout (or cancellation), we see the above log.

Testing server:

  *   OS: Ubuntu 18.04
  *   Julia version: 1.5, 1.6, 1.7
  *   OpticSim.jl version: latest (0.3.1 + new emitter code)

-

**Comment by BrianGun** [comment]
no idea why this might be happening.

**Comment by friggog** [comment]
I had this problem before but I can’t remember if there was any solution. I think at the time it may have just been to avoid calling whatever function results in fort caching…
A little weird why this doesn’t happen on both windows and linux though?

I think if you can cache (some of) the pipeline stages as we did on devops (specifically the .Julia folder) then it only needs to happen once and after that you’re good, maybe just extend the timeout to allow this - although 6 hours should do it so maybe it’s something else…

Take a look at other package’s CI which makes use of Makie and if not raise an issue with Makie is my advice

**Comment by alfredclwong** [comment]
Even weirder, when I exclude the new emitter Vis code, Julia 1.7 gets through fine on linux.

We're doing some caching in CI.yml, but it's limited to .julia/artifacts. Extending to .julia/ could be a good idea, but as you say 6 hours is a long time so I'll try to get to the bottom of this first.

Need to do some more digging - I'll raise an issue when I can actually say what the issue is 😅

**Comment by alfredclwong** [comment]
Solved in 6733316. I'll keep this issue open until the commit is merged.

Note for future: the error was caused by having multiple Makie backends (GLMakie/WGLMakie in this case). To avoid it, we need to explicitly activate the correct backend using, e.g., GLMakie.activate!().
